### DISCUSSION TITLE
How to optimize my opengraph-image loading times?

### DISCUSSION THREAD

**Discussion by peguerosdc** [vercel/next.js#72006] [Help]
### Summary

Hi! I have my nextjs app deployed in railway.app (it is a node runtime, no edge) and my dynamically generated og images are taking 3 seconds in average to load, which I would like to improve.

Below is a simplified version of my `opengraph-image.tsx` where I am already thinking on some things to improve, but any feedback is appreciated:

- I am loading the fonts from the disk in every request. I took that logic from [vercel's guide](https://vercel.com/docs/functions/og-image-generation), but I'm wondering if there's a more efficient way to do it.
- I am using 3 svgs (I left one as an example) in different parts of the image (I left one as a reference). Do they heavily affect the loading time of the whole image?
- My `<img />` is loading the profile's image from Supabase's storage. I don't think there's a way to optimize this, but I thought it's worth mentioning as it might be affecting as well and maybe someone has any idea.

```javascript
const fs = require("fs").promises;
import { ImageResponse } from "next/og";

export const alt = "Profile image";
export const size = {
  width: 1200,
  height: 630,
};

export const contentType = "image/png";

export default async function og({ params }) {
  const profile = await getProfile(params.id);

  return new ImageResponse(
    (
      <div tw="flex h-full w-full flex-col">
        <div
          style={{
            backgroundImage:
              "url(data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 32 32' width='16' height='16' fill='none'%3e%3ccircle fill='rgb(0 0 0 / 0.2)' id='pattern-circle' cx='10' cy='10' r='1.6257413380501518'%3e%3c/circle%3e%3c/svg%3e)",
          }}
        >
          <img
            src={profile.image}
            alt="Profile picture"
            tw="absolute rounded-full aspect-square h-full w-full"
          />
          <div>
            <h1>{profile.name}</h1>
            <ComponentA profile={profile} />
            <ComponentB profile={profile} />
            <ComponentC profile={profile} />
          </div>
        </div>
      </div>
    ),
    {
      ...size,
      fonts: [
        {
          name: "Body",
          data: await fs.readFile("./assets/Inter-Regular.ttf"),
          style: "normal",
        },
        {
          name: "Heading",
          data: await fs.readFile("./assets/Manrope-Bold.ttf"),
        },
      ],
    },
  );
}

```

### Additional information

```js
next: 14.2.16
node: 22
deployed in: railway.app
```

**Comment by aniruddhaadak80** [comment]
Hi @peguerosdc ,

Most relevant solutions to optimize your OpenGraph image loading times:

1. **Load Fonts Efficiently**:
   - Instead of loading fonts from disk on each request, host the fonts on a CDN or use Google Fonts. This will reduce latency and improve loading times.
   ```javascript
   // Consider using a CDN link for fonts
   fonts: [
     {
       name: "Body",
       data: await fs.readFile("./assets/Inter-Regular.ttf"),
       style: "normal",
     },
     {
       name: "Heading",
       data: await fs.readFile("./assets/Manrope-Bold.ttf"),
     },
   ],
   ```

2. **Optimize SVGs**:
   - Minimize the complexity of your SVG files using tools like SVGO to reduce file sizes. This will help in rendering them faster within your images.
   - If SVGs are not essential, consider using simpler graphics or raster images.

3. **Profile Image Optimization**:
   - Ensure the profile image from Supabase is served in an optimized format (e.g., WebP) and properly resized before being loaded. Use lazy loading for better performance:
   ```javascript
   <img
     src={profile.image}
     alt="Profile picture"
     tw="absolute rounded-full aspect-square h-full w-full"
     loading="lazy" // Enables lazy loading for better performance
   />
   ```

Implementing these strategies will help you significantly enhance the loading times of your OpenGraph images.

Thanks.

**Reply by peguerosdc** [reply]
**About loading fonts efficiently...**
If I download the fonts from a CDN, I will get charged by inbound network usage on every og image that is generated. I would prefer a way to have the fonts cached in my server's memory to load them only once or using `next/font`, but that doesn't seem possible and vercel's suggestion is to [load them from disk](https://vercel.com/guides/using-custom-font).

**About optimizing SVGs**
This actually helped. I replaced them with `base64` strings and it helped, but the biggest improvement came from removing the background's SVG.

**About Profile Image Optimization**
My images are already resized, but using webp is a nice suggestion, thanks! About `loading="lazy"`, isn't that for client-side image loading optimization? The image is generated server-side, so I don't think it would help.

**Reply by aniruddhaadak80** [reply]
Hey @peguerosdc ,

Here’s a brief code adjustment to illustrate how you might cache fonts and consider converting images to WebP format:
```let cachedFonts;

async function loadFonts() {
  if (!cachedFonts) {
    cachedFonts = {
      body: await fs.readFile("./assets/Inter-Regular.ttf"),
      heading: await fs.readFile("./assets/Manrope-Bold.ttf"),
    };
  }
  return cachedFonts;
}

export default async function og({ params }) {
  const profile = await getProfile(params.id);
  await loadFonts();

  return new ImageResponse(
    (
      <div tw="flex h-full w-full flex-col">
        <img src={convertToWebP(profile.image)} alt="Profile picture" />
        <div>
          <h1>{profile.name}</h1>
          {/* Other components */}
        </div>
      </div>
    ),
    {
      /* Image response settings */
    },
  );
}

```

Try it out .

 Thanks .
